File: backend/app/services/recommendations.py
from __future__ import annotations

import calendar
import logging
from datetime import date, datetime, timedelta, timezone
from typing import Any

from app.core.time import market_today
from app.repositories.recommendations import RecommendationsRepository
from app.schemas.recommendations import (
    PriceBar,
    RecommendationEvent,
    RecommendationListData,
    RecommendationOverlayData,
)
from app.services.market_data import MarketDataService

logger = logging.getLogger(__name__)


class RecommendationsService:
    def __init__(self, *, repo: RecommendationsRepository, market_data: MarketDataService):
        self._repo = repo
        self._market_data = market_data

    @staticmethod
    def _normalize_symbol(symbol: str | None) -> str | None:
        s = str(symbol or "").strip().upper()
        return s or None

    @staticmethod
    def _market_today() -> date:
        return market_today()

    @staticmethod
    def _subtract_months(d: date, months: int) -> date:
        m = int(months)
        if m <= 0:
            return d
        month0 = d.month - m
        year = d.year
        while month0 <= 0:
            year -= 1
            month0 += 12
        last_day = calendar.monthrange(year, month0)[1]
        day = min(d.day, last_day)
        return date(year, month0, day)

    @staticmethod
    def _extract_positive_keypoints(summary_obj: Any, *, max_items: int = 6) -> list[str]:
        if not isinstance(summary_obj, dict):
            return []

        items = summary_obj.get("positive")
        if items is None:
            items = summary_obj.get("bull_case")
        if not isinstance(items, list) or not items:
            return []

        out: list[str] = []
        seen: set[str] = set()
        max_items_i = max(1, int(max_items))

        for item in items:
            if len(out) >= max_items_i:
                break

            value: str | None = None
            if isinstance(item, str):
                value = item.strip()
            elif isinstance(item, dict):
                for key in ("claim", "text", "reason", "summary", "content"):
                    raw = item.get(key)
                    if isinstance(raw, str) and raw.strip():
                        value = raw.strip()
                        break
            else:
                raw = str(item).strip()
                value = raw or None

            if not value:
                continue
            if value.lower() in {"none", "null", "{}", "[]"}:
                continue
            if value in seen:
                continue
            seen.add(value)
            out.append(value)

        return out

    def list_recommendations(self, *, symbol: str | None, days: int, limit: int) -> RecommendationListData:
        sym = self._normalize_symbol(symbol)
        days = max(1, int(days))
        limit = max(1, int(limit))

        rows = self._repo.fetch_recommendation_rows(symbol=sym, days=days, limit=limit)
        summary_rows = self._repo.fetch_summary_rows_for_recommendations(
            video_ids=[str(r.get("video_id") or "").strip() for r in rows],
            tickers=[str(r.get("ticker") or "").strip() for r in rows],
        )

        positive_keypoints_by_pair: dict[tuple[str, str], list[str]] = {}
        for row in summary_rows:
            vid = str(row.get("video_id") or "").strip()
            ticker = self._normalize_symbol(row.get("ticker"))
            if not vid or not ticker:
                continue
            positive_keypoints_by_pair[(vid, ticker)] = self._extract_positive_keypoints(row.get("summary"))

        events: list[RecommendationEvent] = []
        for r in rows:
            vid = str(r.get("video_id") or "").strip()
            ticker = self._normalize_symbol(r.get("ticker"))
            if not vid or not ticker or ticker == "MARKET":
                continue

            v = r.get("videos")
            if isinstance(v, list):
                v = v[0] if v else None
            if not isinstance(v, dict):
                v = {}

            published_at = r.get("published_at") or v.get("published_at")

            events.append(
                RecommendationEvent(
                    video_id=vid,
                    ticker=ticker,
                    action="buy",
                    title=v.get("title"),
                    channel=v.get("channel"),
                    published_at=published_at,
                    video_url=v.get("video_url"),
                    thumbnail_url=v.get("thumbnail_url"),
                    positive_keypoints=positive_keypoints_by_pair.get((vid, ticker), []),
                )
            )

        return RecommendationListData(items=events)

    @staticmethod
    def _close_on_or_after(
        prices: list[PriceBar], target: date, *, fallback_to_latest: bool = True
    ) -> tuple[str | None, float | None]:
        if not prices:
            return None, None

        def _px(b: PriceBar) -> float | None:
            if b.adj_close is not None:
                try:
                    return float(b.adj_close)
                except Exception:
                    return None
            if b.close is None:
                return None
            try:
                return float(b.close)
            except Exception:
                return None

        parsed: list[tuple[date, str, float | None]] = []
        for b in prices:
            d = str(b.date or "").strip()
            if not d:
                continue
            try:
                bd = date.fromisoformat(d)
            except Exception:
                continue
            parsed.append((bd, d, _px(b)))

        if not parsed:
            return None, None

        candidate: tuple[date, str, float | None] | None = None
        for row in parsed:
            bd = row[0]
            if bd < target:
                continue
            if candidate is None or bd < candidate[0]:
                candidate = row

        if candidate is not None:
            _, d, px = candidate
            return d, px

        if not fallback_to_latest:
            return None, None

        latest = max(parsed, key=lambda r: r[0])
        _, d, px = latest
        return d, px

    @staticmethod
    def _compute_return(entry_close: float | None, exit_close: float | None) -> float | None:
        if entry_close is None or exit_close is None:
            return None
        try:
            entry_f = float(entry_close)
            exit_f = float(exit_close)
        except Exception:
            return None
        if entry_f <= 0:
            return None
        return (exit_f - entry_f) / entry_f

    def get_recommendation_overlay(self, *, symbol: str, days: int) -> RecommendationOverlayData:
        sym = self._normalize_symbol(symbol)
        if not sym:
            return RecommendationOverlayData(symbol="", prices=[], events=[])

        days_i = max(1, int(days))

        recs = self.list_recommendations(symbol=sym, days=days_i, limit=2000)
        events = recs.items
        if not events:
            return RecommendationOverlayData(symbol=sym, prices=[], events=[])

        end = self._market_today()
        start = end - timedelta(days=days_i)

        prices_raw: list[dict[str, Any]]
        try:
            prices_raw = self._market_data.fetch_daily_close_series(symbol=sym, start=start, end=end)
        except Exception:
            logger.warning("market data unavailable for overlay", extra={"symbol": sym})
            prices_raw = []

        prices: list[PriceBar] = []
        for b in prices_raw:
            if not isinstance(b, dict):
                continue
            try:
                prices.append(PriceBar.model_validate(b))
            except Exception:
                continue

        latest_date, latest_close = self._close_on_or_after(prices, end)

        enriched: list[RecommendationEvent] = []
        for e in events:
            pub = e.published_at
            pub_dt: datetime | None = None
            if isinstance(pub, str) and pub.strip():
                try:
                    pub_dt = datetime.fromisoformat(pub.replace("Z", "+00:00")).astimezone(timezone.utc)
                except Exception:
                    pub_dt = None
            elif isinstance(pub, datetime):
                try:
                    pub_dt = pub.astimezone(timezone.utc)
                except Exception:
                    pub_dt = None

            # Event day: use market date (ET) when possible, but keep existing behavior forgiving.
            event_day = (pub_dt.date() if pub_dt is not None else None) or end

            entry_date, entry_close = self._close_on_or_after(prices, event_day)
            if entry_date is None or entry_close is None:
                entry_date, entry_close = latest_date, latest_close

            day_7_date, day_7_close = self._close_on_or_after(
                prices, event_day + timedelta(days=7), fallback_to_latest=False
            )
            day_30_date, day_30_close = self._close_on_or_after(
                prices, event_day + timedelta(days=30), fallback_to_latest=False
            )

            ret = self._compute_return(entry_close, latest_close)
            ret_7d = self._compute_return(entry_close, day_7_close)
            ret_30d = self._compute_return(entry_close, day_30_close)

            payload = e.model_dump(
                exclude={
                    "entry_date",
                    "entry_close",
                    "latest_date",
                    "latest_close",
                    "return_pct",
                    "return_7d_pct",
                    "return_30d_pct",
                }
            )
            payload["entry_date"] = entry_date
            payload["entry_close"] = entry_close
            payload["latest_date"] = latest_date
            payload["latest_close"] = latest_close
            payload["return_pct"] = ret
            payload["return_7d_pct"] = ret_7d
            payload["return_30d_pct"] = ret_30d

            enriched.append(
                RecommendationEvent(
                    **payload,
                )
            )

        return RecommendationOverlayData(symbol=sym, prices=prices, events=enriched)

File: backend/internal/svc/recommendations.go
package svc

import (
	"time"

	"yunews/backend/internal/apperr"
	"yunews/backend/internal/repo"
	"yunews/backend/internal/supabase"
	"yunews/backend/internal/timeutil"
)

// RecommendationEvent mirrors recommendations schema.
type RecommendationEvent struct {
	VideoID           string   `json:"video_id"`
	Ticker            string   `json:"ticker"`
	Action            string   `json:"action"`
	Title             *string  `json:"title,omitempty"`
	Channel           *string  `json:"channel,omitempty"`
	PublishedAt       *string  `json:"published_at,omitempty"`
	VideoURL          *string  `json:"video_url,omitempty"`
	ThumbnailURL      *string  `json:"thumbnail_url,omitempty"`
	PositiveKeypoints []string `json:"positive_keypoints"`
	EntryDate         *string  `json:"entry_date,omitempty"`
	EntryClose        *float64 `json:"entry_close,omitempty"`
	LatestDate        *string  `json:"latest_date,omitempty"`
	LatestClose       *float64 `json:"latest_close,omitempty"`
	ReturnPct         *float64 `json:"return_pct,omitempty"`
	Return7dPct       *float64 `json:"return_7d_pct,omitempty"`
	Return30dPct      *float64 `json:"return_30d_pct,omitempty"`
}

// RecommendationListData is the response for the list endpoint.
type RecommendationListData struct {
	Items []RecommendationEvent `json:"items"`
}

// RecommendationOverlayData is the response for the overlay endpoint.
type RecommendationOverlayData struct {
	Symbol string                `json:"symbol"`
	Prices []repo.CloseBar       `json:"prices"`
	Events []RecommendationEvent `json:"events"`
}

// RecommendationsService handles recommendation business logic.
type RecommendationsService struct {
	repo        *repo.RecommendationsRepository
	marketData  *MarketDataService
}

// NewRecommendationsService creates a RecommendationsService.
func NewRecommendationsService(r *repo.RecommendationsRepository, mds *MarketDataService) *RecommendationsService {
	return &RecommendationsService{repo: r, marketData: mds}
}

// ListRecommendations returns a filtered, shaped list of recommendation events.
func (s *RecommendationsService) ListRecommendations(symbol *string, days, limit int) (*RecommendationListData, error) {
	if days < 1 {
		days = 1
	}
	if limit < 1 {
		limit = 1
	}

	start := timeutil.MarketToday().AddDate(0, 0, -days)
	startISO := start.Format("2006-01-02")

	var sym string
	if symbol != nil {
		sym = *symbol
	}

	rows, err := s.repo.FetchRecommendationRows(sym, startISO, limit)
	if err != nil {
		return nil, err
	}

	// Collect video_ids and tickers for the summary query.
	videoIDs := make([]string, 0, len(rows))
	tickers := make([]string, 0, len(rows))
	for _, r := range rows {
		vid := supabase.GetStr(r, "video_id")
		ticker := normalizeSymbol(supabase.GetStr(r, "ticker"))
		if vid != "" {
			videoIDs = append(videoIDs, vid)
		}
		if ticker != "" {
			tickers = append(tickers, ticker)
		}
	}

	summaryRows, err := s.repo.FetchSummaryRowsForRecommendations(videoIDs, tickers)
	if err != nil {
		return nil, err
	}

	// Map (video_id, ticker) → positive keypoints.
	type pairKey struct{ vid, ticker string }
	keypointsMap := map[pairKey][]string{}
	for _, sr := range summaryRows {
		vid := supabase.GetStr(sr, "video_id")
		ticker := normalizeSymbol(supabase.GetStr(sr, "ticker"))
		if vid == "" || ticker == "" {
			continue
		}
		summaryRaw := sr["summary"]
		keypointsMap[pairKey{vid, ticker}] = extractPositiveKeypoints(summaryRaw, 6)
	}

	events := make([]RecommendationEvent, 0, len(rows))
	for _, r := range rows {
		vid := supabase.GetStr(r, "video_id")
		ticker := normalizeSymbol(supabase.GetStr(r, "ticker"))
		if vid == "" || ticker == "" || ticker == "MARKET" {
			continue
		}

		// videos is a nested object (first element if array).
		v := supabase.GetMap(r, "videos")

		publishedAt := supabase.GetStrPtr(r, "published_at")
		if publishedAt == nil {
			if pa := supabase.GetStrPtr(v, "published_at"); pa != nil {
				publishedAt = pa
			}
		}

		events = append(events, RecommendationEvent{
			VideoID:           vid,
			Ticker:            ticker,
			Action:            "buy",
			Title:             supabase.GetStrPtr(v, "title"),
			Channel:           supabase.GetStrPtr(v, "channel"),
			PublishedAt:       publishedAt,
			VideoURL:          supabase.GetStrPtr(v, "video_url"),
			ThumbnailURL:      supabase.GetStrPtr(v, "thumbnail_url"),
			PositiveKeypoints: keypointsMap[pairKey{vid, ticker}],
		})
	}

	return &RecommendationListData{Items: events}, nil
}

// GetRecommendationOverlay returns price series + enriched recommendation events.
func (s *RecommendationsService) GetRecommendationOverlay(symbol string, days int) (*RecommendationOverlayData, error) {
	sym := normalizeSymbol(symbol)
	if sym == "" {
		return nil, apperr.BadRequest("symbol is required")
	}
	if days < 1 {
		days = 1
	}

	recs, err := s.ListRecommendations(&sym, days, 2000)
	if err != nil {
		return nil, err
	}
	if len(recs.Items) == 0 {
		return &RecommendationOverlayData{Symbol: sym, Prices: []repo.CloseBar{}, Events: []RecommendationEvent{}}, nil
	}

	end := timeutil.MarketToday()
	start := end.AddDate(0, 0, -days)

	prices, _ := s.marketData.FetchDailyCloseSeries(sym, start, end)
	if prices == nil {
		prices = []repo.CloseBar{}
	}

	// Determine latest available close.
	latestDate, latestClose := closeOnOrAfter(prices, end, true)

	enriched := make([]RecommendationEvent, 0, len(recs.Items))
	for _, e := range recs.Items {
		// Determine event day from published_at.
		eventDay := end
		if e.PublishedAt != nil && *e.PublishedAt != "" {
			if t, err2 := time.Parse(time.RFC3339, *e.PublishedAt); err2 == nil {
				ny := timeutil.MarketTZ
				eventDay = t.In(ny).Truncate(24 * time.Hour).In(ny).Round(0)
				// Convert to a plain date.
				eventDay = time.Date(eventDay.Year(), eventDay.Month(), eventDay.Day(), 0, 0, 0, 0, time.UTC)
			}
		}

		entryDate, entryClose := closeOnOrAfter(prices, eventDay, true)
		if entryDate == "" {
			entryDate = latestDate
			entryClose = latestClose
		}

		day7Date, day7Close := closeOnOrAfter(prices, eventDay.AddDate(0, 0, 7), false)
		day30Date, day30Close := closeOnOrAfter(prices, eventDay.AddDate(0, 0, 30), false)
		_ = day7Date
		_ = day30Date

		ret := computeReturn(entryClose, latestClose)
		ret7d := computeReturn(entryClose, day7Close)
		ret30d := computeReturn(entryClose, day30Close)

		ev := e // copy
		ev.EntryDate = strPtr(entryDate)
		ev.EntryClose = entryClose
		ev.LatestDate = strPtr(latestDate)
		ev.LatestClose = latestClose
		ev.ReturnPct = ret
		ev.Return7dPct = ret7d
		ev.Return30dPct = ret30d

		enriched = append(enriched, ev)
	}

	return &RecommendationOverlayData{Symbol: sym, Prices: prices, Events: enriched}, nil
}

// closeOnOrAfter returns the first price bar on or after target, or latest if fallback is true.
func closeOnOrAfter(prices []repo.CloseBar, target time.Time, fallback bool) (string, *float64) {
	if len(prices) == 0 {
		return "", nil
	}

	targetDate := target.UTC().Format("2006-01-02")

	var bestDate string
	var bestClose *float64

	var latestDate string
	var latestClose *float64

	for _, b := range prices {
		d := b.Date
		if d == "" {
			continue
		}

		cl := chooseClose(&b)

		// Track latest overall.
		if latestDate == "" || d > latestDate {
			latestDate = d
			latestClose = cl
		}

		// Track first on-or-after target.
		if d >= targetDate {
			if bestDate == "" || d < bestDate {
				bestDate = d
				bestClose = cl
			}
		}
	}

	if bestDate != "" {
		return bestDate, bestClose
	}
	if fallback {
		return latestDate, latestClose
	}
	return "", nil
}

func chooseClose(b *repo.CloseBar) *float64 {
	if b.AdjClose != nil {
		return b.AdjClose
	}
	return b.Close
}

func computeReturn(entry, exit *float64) *float64 {
	if entry == nil || exit == nil || *entry <= 0 {
		return nil
	}
	r := (*exit - *entry) / *entry
	return &r
}

func extractPositiveKeypoints(summaryRaw interface{}, maxItems int) []string {
	m, ok := summaryRaw.(map[string]interface{})
	if !ok {
		return nil
	}

	var items []interface{}
	if v, ok2 := m["positive"]; ok2 {
		if sl, ok3 := v.([]interface{}); ok3 {
			items = sl
		}
	}
	if items == nil {
		if v, ok2 := m["bull_case"]; ok2 {
			if sl, ok3 := v.([]interface{}); ok3 {
				items = sl
			}
		}
	}
	if len(items) == 0 {
		return nil
	}

	seen := map[string]bool{}
	out := make([]string, 0, maxItems)
	for _, item := range items {
		if len(out) >= maxItems {
			break
		}
		var value string
		switch v := item.(type) {
		case string:
			value = v
		case map[string]interface{}:
			for _, k := range []string{"claim", "text", "reason", "summary", "content"} {
				if raw, ok2 := v[k]; ok2 {
					if s, ok3 := raw.(string); ok3 && s != "" {
						value = s
						break
					}
				}
			}
		}
		if value == "" || seen[value] {
			continue
		}
		seen[value] = true
		out = append(out, value)
	}
	return out
}

func strPtr(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}
